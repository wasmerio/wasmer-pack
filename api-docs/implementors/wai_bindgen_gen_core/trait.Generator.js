(function() {var implementors = {
"wai_bindgen_gen_js":[["impl <a class=\"trait\" href=\"wai_bindgen_gen_core/trait.Generator.html\" title=\"trait wai_bindgen_gen_core::Generator\">Generator</a> for <a class=\"struct\" href=\"wai_bindgen_gen_js/struct.Js.html\" title=\"struct wai_bindgen_gen_js::Js\">Js</a>"]],
"wai_bindgen_gen_rust_wasm":[["impl Generator for <a class=\"struct\" href=\"wai_bindgen_gen_rust_wasm/struct.RustWasm.html\" title=\"struct wai_bindgen_gen_rust_wasm::RustWasm\">RustWasm</a>"]],
"wai_bindgen_gen_wasmer_py":[["impl <a class=\"trait\" href=\"wai_bindgen_gen_core/trait.Generator.html\" title=\"trait wai_bindgen_gen_core::Generator\">Generator</a> for <a class=\"struct\" href=\"wai_bindgen_gen_wasmer_py/struct.WasmerPy.html\" title=\"struct wai_bindgen_gen_wasmer_py::WasmerPy\">WasmerPy</a>"]]
};if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()