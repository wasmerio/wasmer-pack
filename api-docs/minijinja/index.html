<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="MiniJinja is a powerful but minimal dependency template engine for Rust which is based on the syntax and behavior of the Jinja2 template engine for Python.  It’s implemented on top of `serde`.  The goal is to be able to render a large chunk of the Jinja2 template ecosystem from Rust with a minimal engine and to leverage an already existing ecosystem of editor integrations."><title>minijinja - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-ba5701c5741a7b69.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="minijinja" data-themes="" data-resource-suffix="" data-rustdoc-version="1.70.0 (90c541806 2023-05-31)" data-search-js="search-e077946657036a58.js" data-settings-js="settings-298e1ea74db45b39.js" data-settings-css="settings-7bfb4c59cc6bc502.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f61008743c98d196.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../minijinja/index.html"><img src="https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo-square.png" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../minijinja/index.html"><img src="https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo-square.png" alt="logo"></a><h2 class="location"><a href="#">Crate minijinja</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.34.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#functions">Functions</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">minijinja</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/minijinja/lib.rs.html#1-260">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><div align=center>
  <img src="https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo.png" alt="" width=320>
  <p><strong>MiniJinja: a powerful template engine for Rust with minimal dependencies</strong></p>
</div>
<p>MiniJinja is a powerful but minimal dependency template engine for Rust which
is based on the syntax and behavior of the
<a href="https://jinja.palletsprojects.com/">Jinja2</a> template engine for Python.  It’s
implemented on top of <a href="../serde/index.html" title="mod serde"><code>serde</code></a>.  The goal is to be able to render a large
chunk of the Jinja2 template ecosystem from Rust with a minimal engine and to
leverage an already existing ecosystem of editor integrations.</p>
<div class="example-wrap"><pre class="language-jinja"><code>{% for user in users %}
  &lt;li&gt;{{ user.name }}&lt;/li&gt;
{% endfor %}
</code></pre></div>
<p>You can play with MiniJinja online <a href="https://mitsuhiko.github.io/minijinja-playground/">in the browser
playground</a> powered by a
WASM build of MiniJinja.</p>
<h2 id="why-minijinja"><a href="#why-minijinja">Why MiniJinja</a></h2>
<p>MiniJinja by its name wants to be a good default choice if you need a little
bit of templating with minimal dependencies.  It has the following goals:</p>
<ul>
<li>Well documented, compact API</li>
<li>Minimal dependencies, reasonable compile times and decent runtime performance</li>
<li>Stay close as possible to Jinja2</li>
<li>Support for expression evaluation</li>
<li>Support for all <code>serde</code> compatible types</li>
<li>Excellent test coverage</li>
<li>Support for dynamic runtime objects with methods and dynamic attributes</li>
</ul>
<h2 id="template-usage"><a href="#template-usage">Template Usage</a></h2>
<p>To use MiniJinja one needs to create an <a href="struct.Environment.html" title="struct minijinja::Environment"><code>Environment</code></a> and populate it with
templates.  Afterwards templates can be loaded and rendered.  To pass data
one can pass any serde serializable value.  The <a href="macro.context.html" title="macro minijinja::context"><code>context!</code></a> macro can be
used to quickly construct a template context:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>minijinja::{Environment, context};

<span class="kw">let </span><span class="kw-2">mut </span>env = Environment::new();
env.add_template(<span class="string">&quot;hello&quot;</span>, <span class="string">&quot;Hello {{ name }}!&quot;</span>).unwrap();
<span class="kw">let </span>tmpl = env.get_template(<span class="string">&quot;hello&quot;</span>).unwrap();
<span class="macro">println!</span>(<span class="string">&quot;{}&quot;</span>, tmpl.render(<span class="macro">context!</span>(name =&gt; <span class="string">&quot;John&quot;</span>)).unwrap());</code></pre></div>
<div class="example-wrap"><pre class="language-plain"><code>Hello John!
</code></pre></div>
<p>For super trivial cases where you need to render a string once, you can
also use the <a href="macro.render.html" title="macro minijinja::render"><code>render!</code></a> macro which acts a bit like a replacement
for the <a href="https://doc.rust-lang.org/1.70.0/alloc/macro.format.html" title="macro alloc::format"><code>format!</code></a> macro.</p>
<h2 id="expression-usage"><a href="#expression-usage">Expression Usage</a></h2>
<p>MiniJinja — like Jinja2 — allows to be used as expression language.  This can be
useful to express logic in configuration files or similar things.  For this
purpose the <a href="struct.Environment.html#method.compile_expression" title="method minijinja::Environment::compile_expression"><code>Environment::compile_expression</code></a> method can be used.  It returns
an expression object that can then be evaluated, returning the result:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>minijinja::{Environment, context};

<span class="kw">let </span>env = Environment::new();
<span class="kw">let </span>expr = env.compile_expression(<span class="string">&quot;number &lt; 42&quot;</span>).unwrap();
<span class="kw">let </span>result = expr.eval(<span class="macro">context!</span>(number =&gt; <span class="number">23</span>)).unwrap();
<span class="macro">assert_eq!</span>(result.is_true(), <span class="bool-val">true</span>);</code></pre></div>
<p>This becomes particularly powerful when <a href="value/trait.Object.html" title="trait minijinja::value::Object">dynamic objects</a> are
exposed to templates.</p>
<h2 id="custom-filters"><a href="#custom-filters">Custom Filters</a></h2>
<p>MiniJinja lets you register functions as filter functions (see
<a href="filters/trait.Filter.html" title="trait minijinja::filters::Filter"><code>Filter</code></a>) with the engine.  These can then be
invoked directly from the template:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>minijinja::{Environment, context};

<span class="kw">let </span><span class="kw-2">mut </span>env = Environment::new();
env.add_filter(<span class="string">&quot;repeat&quot;</span>, str::repeat);
env.add_template(<span class="string">&quot;hello&quot;</span>, <span class="string">&quot;{{ &#39;Na &#39;|repeat(3) }} {{ name }}!&quot;</span>).unwrap();
<span class="kw">let </span>tmpl = env.get_template(<span class="string">&quot;hello&quot;</span>).unwrap();
<span class="macro">println!</span>(<span class="string">&quot;{}&quot;</span>, tmpl.render(<span class="macro">context!</span>(name =&gt; <span class="string">&quot;Batman&quot;</span>)).unwrap());</code></pre></div>
<div class="example-wrap"><pre class="language-plain"><code>Na Na Na Batman!
</code></pre></div><h2 id="learn-more"><a href="#learn-more">Learn more</a></h2>
<ul>
<li><a href="struct.Environment.html" title="struct minijinja::Environment"><code>Environment</code></a>: the main API entry point.  Teaches you how to configure the environment.</li>
<li><a href="struct.Template.html" title="struct minijinja::Template"><code>Template</code></a>: the template object API.  Shows you how templates can be rendered.</li>
<li><a href="syntax/index.html" title="mod minijinja::syntax"><code>syntax</code></a>: provides documentation of the template engine syntax.</li>
<li><a href="filters/index.html" title="mod minijinja::filters"><code>filters</code></a>: teaches you how to write custom filters and to see the list of built-in filters.</li>
<li><a href="tests/index.html" title="mod minijinja::tests"><code>tests</code></a>: teaches you how to write custom test functions and to see the list of built-in tests.</li>
<li><a href="functions/index.html" title="mod minijinja::functions"><code>functions</code></a>: teaches how to write custom functions and to see the list of built-in functions.</li>
<li>For auto reloading use the <a href="https://docs.rs/minijinja-autoreload"><code>minijinja-autoreload</code></a> crate.</li>
</ul>
<p>Additionally there is an <a href="https://github.com/mitsuhiko/minijinja/tree/main/examples">list of examples</a>
with many different small example programs on GitHub to explore.</p>
<h2 id="error-handling"><a href="#error-handling">Error Handling</a></h2>
<p>MiniJinja tries to give you good errors out of the box.  However if you use includes or
template inheritance your experience will improve greatly if you ensure to render chained
errors.  For more information see <a href="struct.Error.html" title="struct minijinja::Error"><code>Error</code></a> with an example.</p>
<h2 id="size-and-compile-times"><a href="#size-and-compile-times">Size and Compile Times</a></h2>
<p>MiniJinja attempts to compile fast so it can be used as a sensible template engine choice
when compile times matter.  Because of this it’s internally modular so unnecessary bits and
pieces can be removed.  The default set of features tries to strike a balance but in
situations where only a subset is needed (eg: <code>build.rs</code>) all default features can be
be disabled.</p>
<h2 id="optional-features"><a href="#optional-features">Optional Features</a></h2>
<p>MiniJinja comes with a lot of optional features, some of which are turned on by
default.  If you plan on using MiniJinja in a library, please consider turning
off all default features and to opt-in explicitly into the ones you actually
need.</p>
<details><summary><strong style="cursor: pointer">Configurable Features</strong></summary>
<p>To cut down on size of the engine some default functionality can be removed:</p>
<ul>
<li>
<p><strong>Engine Features:</strong></p>
<ul>
<li><code>builtins</code>: if this feature is removed the default filters, tests and
functions are not implemented.</li>
<li><code>macros</code>: when removed the <code>{% macro %}</code> tag is not included.</li>
<li><code>multi_template</code>: when removed the templates related to imports and extends
are removed (<code>{% from %}</code>, <code>{% import %}</code>, <code>{% include %}</code>, and <code>{% extends %}</code>
as well as <code>{% block %}</code>).</li>
<li><code>adjacent_loop_items</code>: when removed the <code>previtem</code> and <code>nextitem</code> attributes of
the <code>loop</code> object no longer exist.  Removing this feature can provide faster
template execution when a lot of loops are involved.</li>
<li><code>unicode</code>: when added unicode identifiers are supported.  Without this features
only ASCII identifiers can be used for variable names and attributes.</li>
</ul>
</li>
<li>
<p><strong>Rust Functionality:</strong></p>
<ul>
<li><code>debug</code>: if this feature is removed some debug functionality of the engine is
removed as well.  This mainly affects the quality of error reporting.</li>
<li><code>deserialization</code>: when removed this disables deserialization support for
the <a href="value/struct.Value.html" title="struct minijinja::value::Value"><code>Value</code></a> type.</li>
</ul>
</li>
</ul>
<p>There are some additional features that can be enabled:</p>
<ul>
<li><code>fuel</code>: enables the <code>fuel</code> feature which makes the engine track fuel consumption which
can be used to better protect against expensive templates.</li>
<li><code>source</code>: enables the <code>Source</code> type which helps with dynamic loading of templates.</li>
<li><code>speedups</code>: enables all speedups, in particular it turns on the <code>v_htmlescape</code> dependency
for faster HTML escapling.</li>
<li><code>json</code>: When enabled the <code>tojson</code> filter is added as builtin filter as well as
the ability to auto escape via <code>AutoEscape::Json</code>.</li>
<li><code>urlencode</code>: When enabled the <code>urlencode</code> filter is added as builtin filter.</li>
<li><code>preserve_order</code>: When enable the internal value implementation uses an indexmap
which preserves the original order of maps and structs.</li>
<li><code>key_interning</code>: if this feature is enabled the automatic string interning in
the value type is enabled.  This feature used to be turned on by default but
has negative performance effects in newer versions of MiniJinja since a lot of
the previous uses of key interning are no longer needed.  Enabling it however
cuts down on memory usage slightly in certain scenarios by interning all string
keys used in dynamic map values.</li>
<li><code>custom_syntax</code>: when this feature is enabled, custom delimiters are supported by
the parser.</li>
</ul>
</details></div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="filters/index.html" title="mod minijinja::filters">filters</a></div><div class="desc docblock-short">Filter functions and abstractions.</div></li><li><div class="item-name"><a class="mod" href="functions/index.html" title="mod minijinja::functions">functions</a></div><div class="desc docblock-short">Global functions and abstractions.</div></li><li><div class="item-name"><a class="mod" href="syntax/index.html" title="mod minijinja::syntax">syntax</a></div><div class="desc docblock-short">Documents the syntax for templates.</div></li><li><div class="item-name"><a class="mod" href="tests/index.html" title="mod minijinja::tests">tests</a></div><div class="desc docblock-short">Test functions and abstractions.</div></li><li><div class="item-name"><a class="mod" href="value/index.html" title="mod minijinja::value">value</a></div><div class="desc docblock-short">Provides a dynamic value type abstraction.</div></li></ul><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.context.html" title="macro minijinja::context">context</a></div><div class="desc docblock-short">Creates a template context with keys and values.</div></li><li><div class="item-name"><a class="macro" href="macro.render.html" title="macro minijinja::render">render</a></div><div class="desc docblock-short">A macro similar to <a href="https://doc.rust-lang.org/1.70.0/alloc/macro.format.html" title="macro alloc::format"><code>format!</code></a> but that uses MiniJinja for rendering.</div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Environment.html" title="struct minijinja::Environment">Environment</a></div><div class="desc docblock-short">An abstraction that holds the engine configuration.</div></li><li><div class="item-name"><a class="struct" href="struct.Error.html" title="struct minijinja::Error">Error</a></div><div class="desc docblock-short">Represents template errors.</div></li><li><div class="item-name"><a class="struct" href="struct.Expression.html" title="struct minijinja::Expression">Expression</a></div><div class="desc docblock-short">A handle to a compiled expression.</div></li><li><div class="item-name"><a class="struct" href="struct.HtmlEscape.html" title="struct minijinja::HtmlEscape">HtmlEscape</a></div><div class="desc docblock-short">Helper to HTML escape a string.</div></li><li><div class="item-name"><a class="struct" href="struct.Output.html" title="struct minijinja::Output">Output</a></div><div class="desc docblock-short">An abstraction over <a href="https://doc.rust-lang.org/1.70.0/core/fmt/trait.Write.html" title="trait core::fmt::Write"><code>Write</code></a> for the rendering.</div></li><li><div class="item-name"><a class="struct" href="struct.State.html" title="struct minijinja::State">State</a></div><div class="desc docblock-short">Provides access to the current execution state of the engine.</div></li><li><div class="item-name"><a class="struct" href="struct.Template.html" title="struct minijinja::Template">Template</a></div><div class="desc docblock-short">Represents a handle to a template.</div></li></ul><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.AutoEscape.html" title="enum minijinja::AutoEscape">AutoEscape</a></div><div class="desc docblock-short">Controls the autoescaping behavior.</div></li><li><div class="item-name"><a class="enum" href="enum.ErrorKind.html" title="enum minijinja::ErrorKind">ErrorKind</a></div><div class="desc docblock-short">An enum describing the error kind.</div></li><li><div class="item-name"><a class="enum" href="enum.UndefinedBehavior.html" title="enum minijinja::UndefinedBehavior">UndefinedBehavior</a></div><div class="desc docblock-short">Defines the behavior of undefined values in the engine.</div></li></ul><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.default_auto_escape_callback.html" title="fn minijinja::default_auto_escape_callback">default_auto_escape_callback</a></div><div class="desc docblock-short">The default logic for auto escaping based on file extension.</div></li><li><div class="item-name"><a class="fn" href="fn.escape_formatter.html" title="fn minijinja::escape_formatter">escape_formatter</a></div><div class="desc docblock-short">The default formatter.</div></li></ul></section></div></main></body></html>