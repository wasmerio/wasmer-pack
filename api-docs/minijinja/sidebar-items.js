window.SIDEBAR_ITEMS = {"enum":["AutoEscape","ErrorKind","UndefinedBehavior"],"fn":["default_auto_escape_callback","escape_formatter"],"macro":["context","render"],"mod":["filters","functions","syntax","tests","value"],"struct":["Environment","Error","Expression","HtmlEscape","Output","State","Template"]};