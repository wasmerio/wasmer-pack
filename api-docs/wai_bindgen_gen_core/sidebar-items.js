window.SIDEBAR_ITEMS = {"enum":["Direction"],"fn":["load"],"macro":["uwrite","uwriteln"],"struct":["Files","Ns","Source","TypeInfo","Types"],"trait":["Generator"]};