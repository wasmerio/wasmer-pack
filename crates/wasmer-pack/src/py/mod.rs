use std::path::Path;

use anyhow::Error;
use heck::{ToPascalCase, ToSnakeCase};
use minijinja::Environment;
use once_cell::sync::Lazy;
use wai_bindgen_gen_core::Generator;
use wai_bindgen_gen_wasmer_py::WasmerPy;

use crate::{
    types::{BindingsOptions, Interface, Package},
    Files, Metadata, Module, SourceFile,
};

static TEMPLATES: Lazy<Environment> = Lazy::new(|| {
    compile_templates!(
        "bindings.__init__.py",
        "top_level.__init__.py",
        "MANIFEST.in",
        "commands.__init__.py",
    )
});

/// Generate Python bindings.
pub fn generate_python(package: &Package, options: &BindingsOptions) -> Result<Files, Error> {
    let metadata = package.metadata();

    // make sure the name is in snake-case
    let package_name = if let Some(name) = &options.name {
        name.to_snake_case()
    } else {
        metadata.package_name.name().to_string().to_snake_case()
    };

    let mut files = Files::new();

    let ctx = Context::for_package(package);

    if !ctx.libraries.is_empty() {
        files.insert_child_directory(
            Path::new(&package_name).join("bindings"),
            library_bindings(&ctx)?,
        );
    }

    if !ctx.commands.is_empty() {
        files.insert_child_directory(
            Path::new(&package_name).join("commands"),
            command_bindings(&ctx)?,
        );
    }

    files.insert(
        Path::new(&package_name).join("__init__.py"),
        top_level_dunder_init(package)?,
    );
    // Indicate that we use type hints
    files.insert(
        Path::new(&package_name).join("py.typed"),
        SourceFile::empty(),
    );

    files.insert(
        "pyproject.toml",
        generate_pyproject_toml(metadata, &package_name)?,
    );

    files.insert("MANIFEST.in", generate_manifest(package, &package_name)?);

    Ok(files)
}

#[derive(Debug, serde::Serialize)]
struct Context {
    commands: Vec<CommandContext>,
    libraries: Vec<LibraryContext>,
}

impl Context {
    fn for_package(pkg: &Package) -> Self {
        let commands = pkg
            .commands()
            .iter()
            .cloned()
            .map(CommandContext::from)
            .collect();

        let libraries = pkg
            .libraries()
            .iter()
            .cloned()
            .map(LibraryContext::from)
            .collect();

        Context {
            commands,
            libraries,
        }
    }
}

#[derive(Debug, serde::Serialize)]
struct LibraryContext {
    ident: String,
    class_name: String,
    module_filename: String,
    wasi: bool,
    exports: InterfaceContext,
    imports: Vec<InterfaceContext>,
    #[serde(skip)]
    module: Module,
}

impl From<crate::Library> for LibraryContext {
    fn from(lib: crate::Library) -> Self {
        let module_filename = Path::new(lib.module_filename()).with_extension("wasm");
        let ident = lib.interface_name().to_snake_case();
        let class_name = lib.class_name();

        LibraryContext {
            ident,
            class_name,
            module_filename: module_filename.display().to_string(),
            wasi: lib.requires_wasi(),
            exports: lib.exports.into(),
            imports: lib
                .imports
                .into_iter()
                .map(InterfaceContext::from)
                .collect(),
            module: lib.module,
        }
    }
}

#[derive(Debug, serde::Serialize)]
struct InterfaceContext {
    /// The name used when you need to refer to this interface as a variable.
    ident: String,
    /// The name of the interface (i.e. the `wasmer-pack` in
    /// `wasmer-pack.exports.wit`).
    interface_name: String,
    /// The name of the class generated by `wai-bindgen` (i.e. `WasmerPack`).
    class_name: String,
    #[serde(skip)]
    interface: Interface,
}

impl From<Interface> for InterfaceContext {
    fn from(interface: Interface) -> Self {
        InterfaceContext {
            ident: interface.name().to_snake_case(),
            interface_name: interface.name().to_string(),
            class_name: interface.name().to_pascal_case(),
            interface,
        }
    }
}

#[derive(Debug, serde::Serialize)]
struct CommandContext {
    name: String,
    ident: String,
    module_filename: String,
    #[serde(skip)]
    wasm: Vec<u8>,
}

impl From<crate::Command> for CommandContext {
    fn from(cmd: crate::Command) -> CommandContext {
        let ident = cmd.name.to_snake_case();
        let module_filename = Path::new(&ident).with_extension("wasm");
        CommandContext {
            name: cmd.name.clone(),
            ident,
            module_filename: module_filename.display().to_string(),
            wasm: cmd.wasm,
        }
    }
}

fn command_bindings(ctx: &Context) -> Result<Files, Error> {
    let mut files = Files::new();

    for cmd in &ctx.commands {
        files.insert(&cmd.module_filename, SourceFile::from(&cmd.wasm));
    }

    files.insert(
        "__init__.py",
        TEMPLATES
            .get_template("commands.__init__.py")
            .unwrap()
            .render(ctx)?
            .into(),
    );

    Ok(files)
}

fn library_bindings(ctx: &Context) -> Result<Files, Error> {
    let mut files = Files::new();

    for lib in &ctx.libraries {
        let mut bindings = generate_bindings(lib);
        bindings.insert(&lib.module_filename, lib.module.wasm.clone().into());
        files.insert_child_directory(&lib.ident, bindings);
    }

    let dunder_init = TEMPLATES
        .get_template("bindings.__init__.py")
        .unwrap()
        .render(ctx)?;
    files.insert("__init__.py", dunder_init.into());

    Ok(files)
}

fn generate_manifest(package: &Package, package_name: &str) -> Result<SourceFile, Error> {
    let ctx = minijinja::context! {
        package_name,
        libraries => package.libraries()
        .iter()
            .map(|lib| lib.interface_name())
            .collect::<Vec<_>>(),
        commands => package.commands()
            .iter()
            .map(|cmd| cmd.name.as_str())
            .collect::<Vec<_>>(),
    };
    let rendered = TEMPLATES
        .get_template("MANIFEST.in")
        .unwrap()
        .render(&ctx)?;

    Ok(rendered.into())
}

fn generate_pyproject_toml(metadata: &Metadata, package_name: &str) -> Result<SourceFile, Error> {
    let Metadata {
        version,
        description,
        ..
    } = metadata;

    let project = PyProject {
        project: Project {
            name: package_name,
            version,
            description: description.as_deref(),
            readme: None,
            keywords: Vec::new(),
            dependencies: vec!["wasmer", "wasmer_compiler_cranelift"],
        },
        build_system: BuildSystem {
            requires: &["setuptools", "setuptools-scm"],
            build_backend: "setuptools.build_meta",
        },
    };

    let serialized = toml::to_string(&project)?;

    Ok(serialized.into())
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "kebab-case")]
struct PyProject<'a> {
    project: Project<'a>,
    build_system: BuildSystem<'a>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "kebab-case")]
struct BuildSystem<'a> {
    requires: &'a [&'a str],
    build_backend: &'a str,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
struct Project<'a> {
    name: &'a str,
    version: &'a str,
    description: Option<&'a str>,
    readme: Option<&'a Path>,
    keywords: Vec<&'a str>,
    dependencies: Vec<&'a str>,
}

fn top_level_dunder_init(package: &Package) -> Result<SourceFile, Error> {
    let Metadata {
        version,
        description,
        package_name,
    } = package.metadata();

    let ctx = minijinja::context! {
        version,
        description,
        generator => crate::GENERATOR,
        package_name => package_name.to_string(),
        ident => package_name.name().to_pascal_case(),
        commands => !package.commands().is_empty(),
        libraries => !package.libraries().is_empty(),
    };

    let rendered = TEMPLATES
        .get_template("top_level.__init__.py")
        .unwrap()
        .render(ctx)?;

    Ok(rendered.into())
}

fn generate_bindings(lib: &LibraryContext) -> Files {
    // Note: imports and exports were reported from the perspective of the
    // guest, but we're generating bindings from the perspective of the host.
    // Hence the "host_imports = guest_exports" thing.

    let imports = std::slice::from_ref(&lib.exports.interface.0);
    let exports: Vec<_> = lib
        .imports
        .iter()
        .map(|ctx| ctx.interface.0.clone())
        .collect();

    let mut generated = wai_bindgen_gen_core::Files::default();

    WasmerPy::default().generate_all(imports, &exports, &mut generated);

    let mut files = Files::from(generated);
    files.insert("__init__.py", "from .bindings import *".into());

    files
}

#[cfg(test)]
mod tests {
    use insta::Settings;

    use super::*;
    use crate::{Command, Library, Module};
    use std::collections::BTreeSet;

    const WASMER_PACK_EXPORTS: &str = include_str!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../wasm/wasmer-pack.exports.wai"
    ));

    #[test]
    fn generated_files() {
        let expected: BTreeSet<&Path> = [
            "MANIFEST.in",
            "pyproject.toml",
            "wasmer_pack/__init__.py",
            "wasmer_pack/py.typed",
            "wasmer_pack/commands/__init__.py",
            "wasmer_pack/commands/first.wasm",
            "wasmer_pack/commands/second_with_dashes.wasm",
            "wasmer_pack/bindings/__init__.py",
            "wasmer_pack/bindings/wasmer_pack/__init__.py",
            "wasmer_pack/bindings/wasmer_pack/bindings.py",
            "wasmer_pack/bindings/wasmer_pack/wasmer_pack_wasm.wasm",
        ]
        .iter()
        .map(Path::new)
        .collect();
        let metadata = Metadata::new("wasmer/wasmer-pack".parse().unwrap(), "1.2.3");
        let module = Module {
            name: "wasmer_pack_wasm.wasm".to_string(),
            abi: crate::Abi::None,
            wasm: Vec::new(),
        };
        let exports =
            crate::Interface::from_wit("wasmer-pack.exports.wit", WASMER_PACK_EXPORTS).unwrap();
        let commands = vec![
            Command::new("first", []),
            Command::new("second-with-dashes", []),
        ];
        let browser =
            crate::Interface::from_wit("browser.wit", "greet: func(who: string) -> string")
                .unwrap();
        let libraries = vec![Library {
            module,
            exports,
            imports: vec![browser],
        }];
        let package = Package::new(metadata, libraries, commands);
        let options = BindingsOptions::default();
        let files = generate_python(&package, &options).unwrap();

        let actual_files: BTreeSet<_> = files.iter().map(|(p, _)| p).collect();
        assert_eq!(actual_files, expected);

        let mut settings = Settings::clone_current();
        settings.add_filter(
            r"Generated by wasmer-pack v\d+\.\d+\.\d+(-\w+(\.\d+)?)?",
            "Generated by XXX",
        );
        settings.bind(|| {
            insta::assert_display_snapshot!(files["pyproject.toml"].utf8_contents().unwrap());
            insta::assert_display_snapshot!(files["MANIFEST.in"].utf8_contents().unwrap());
            insta::assert_display_snapshot!(files["wasmer_pack/__init__.py"]
                .utf8_contents()
                .unwrap()
                .replace(crate::GENERATOR, "XXX"));
            insta::assert_display_snapshot!(files["wasmer_pack/bindings/__init__.py"]
                .utf8_contents()
                .unwrap());
            insta::assert_display_snapshot!(files["wasmer_pack/commands/__init__.py"]
                .utf8_contents()
                .unwrap());
        });
        insta::assert_display_snapshot!(files["wasmer_pack/py.typed"].utf8_contents().unwrap());

        let actual_files: BTreeSet<_> = files.iter().map(|(p, _)| p).collect();
        assert_eq!(actual_files, expected);
    }
}
