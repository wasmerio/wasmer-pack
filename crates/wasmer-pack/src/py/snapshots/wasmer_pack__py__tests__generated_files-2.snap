---
source: crates/wasmer-pack/src/py/mod.rs
expression: "files[\"MANIFEST.in\"].utf8_contents().unwrap()"
---
include wasmer_package__wasmer__wasmer_pack/py.typed
include wasmer_package__wasmer__wasmer_pack/bindings/wasmer_pack/*.wasm
include wasmer_package__wasmer__wasmer_pack/commands/*.wasm
